### PR TITLE
Feature/polkadot0.9.16 dev

### PR DESCRIPTION
**Pull request by baidang201**

### CODE REVIEW

File: primitives/Cargo.toml [hunk 1]
```diff
@@ -13,13 +13,13 @@ targets = ["x86_64-unknown-linux-gnu"]
 [dependencies]
 codec = { package = "parity-scale-codec", version = "2.0.0", default-features = false, features = ["derive"] }
```
Comment (xcaptain) [inline primitives/Cargo.toml:14]: other Cargo.toml have upgraded to 2.3.1, would better also change here.

File: runtime/src/lib.rs [hunk 2]
```diff
@@ -159,6 +160,7 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
     impl_version: 0,
     apis: RUNTIME_API_VERSIONS,
     transaction_version: 5,
+    state_version: 1,
```
Comment (xcaptain) [inline runtime/src/lib.rs:163]: use state_version: 0，1 means use new trie structure, may not very stable.

File: runtime/src/lib.rs [hunk 3]
```diff
@@ -383,6 +388,24 @@ impl pallet_scheduler::Config for Runtime {
     type ScheduleOrigin = EnsureRoot<AccountId>;
     type MaxScheduledPerBlock = MaxScheduledPerBlock;
     type WeightInfo = pallet_scheduler::weights::SubstrateWeight<Runtime>;
+    type OriginPrivilegeCmp = EqualPrivilegeOnly;
```
Comment (xcaptain) [inline runtime/src/lib.rs:391]: use the built-in `EqualPrivilegeOnly` or a customized one like polkadot?
```rust
/// Used the compare the privilege of an origin inside the scheduler.
pub struct OriginPrivilegeCmp;

impl PrivilegeCmp<OriginCaller> for OriginPrivilegeCmp {
    fn cmp_privilege(left: &OriginCaller, right: &OriginCaller) -> Option<Ordering> {
        if left == right {
            return Some(Ordering::Equal)
        }

        match (left, right) {
            // Root is greater than anything.
            (OriginCaller::system(frame_system::RawOrigin::Root), _) => Some(Ordering::Greater),
            // Check which one has more yes votes.
            (
                OriginCaller::Council(pallet_collective::RawOrigin::Members(l_yes_votes, l_count)),
                OriginCaller::Council(pallet_collective::RawOrigin::Members(r_yes_votes, r_count)),
            ) => Some((l_yes_votes * r_count).cmp(&(r_yes_votes * l_count))),
            // For every other origin we don't care, as they are not used for `ScheduleOrigin`.
            _ => None,
        }
    }
}
```

File: runtime/src/lib.rs [hunk 4]
```diff
@@ -1594,21 +1633,32 @@ impl_runtime_apis! {
             dest: AccountId,
             value: Balance,
             gas_limit: u64,
+            storage_deposit_limit: Option<Balance>,
```
Comment (xcaptain) [inline runtime/src/lib.rs:1636]: add a nonzero check for `SignedExtra` at line 1382

File: runtime/src/lib.rs [hunk 5]
```diff
@@ -1594,21 +1633,32 @@ impl_runtime_apis! {
...
             input_data: Vec<u8>,
-        ) -> pallet_contracts_primitives::ContractExecResult {
-            Contracts::bare_call(origin, dest, value, gas_limit, input_data, true)
+        ) -> pallet_contracts_primitives::ContractExecResult<Balance> {
+            Contracts::bare_call(origin, dest, value, gas_limit, storage_deposit_limit, input_data, true)
         }
 
         fn instantiate(
             origin: AccountId,
-            endowment: Balance,
```
Comment (xcaptain) [inline runtime/src/lib.rs:1604]: use `AllPalletsWithSystem` other than `AllPallets` in line 1409

File: cli/Cargo.toml [hunk 6]
```diff
@@ -46,104 +46,105 @@ parking_lot = "0.11.1"
...
+sc-cli = { optional = true, git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.16" }
+frame-benchmarking-cli = { optional = true, git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.16" }
 node-inspect = { version = "0.8.0", optional = true, path = "../inspect" }
 
 [target.'cfg(target_arch="x86_64")'.dependencies]
 node-executor = { version = "2.0.0", path = "../executor", features = [ "wasmtime" ] }
-sc-cli = { optional = true, git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.12", features = [ "wasmtime" ] }
-sc-service = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.12", default-features = false, features = [ "wasmtime" ] }
-sp-trie = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.12", default-features = false, features = ["memory-tracker"] }
+sc-cli = { optional = true, git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.16", features = [ "wasmtime" ] }
```
Comment (xcaptain) [inline cli/Cargo.toml:136]: sc-cli, sc-service, sc-executor either all have a `features = [ "wasmtime" ]` or none have this feature.

File: runtime/src/lib.rs (hunk 3 above)
Comment (baidang201) [reply runtime/src/lib.rs:391]: use "type OriginPrivilegeCmp = EqualPrivilegeOnly;" like in  https://github.com/paritytech/substrate/blob/polkadot-v0.9.16/bin/node/runtime/src/lib.rs#L349

File: primitives/Cargo.toml (hunk 1 above)
Comment (baidang201) [reply primitives/Cargo.toml:14]: upgrade to 2.3.1

File: runtime/src/lib.rs (hunk 2 above)
Comment (baidang201) [reply runtime/src/lib.rs:163]: substate 0.9.16 use 1, i think it is stable already.
https://github.com/paritytech/substrate/blob/polkadot-v0.9.16/bin/node/runtime/src/lib.rs#L129

File: runtime/src/lib.rs (hunk 4 above)
Comment (baidang201) [reply runtime/src/lib.rs:1636]: i not sure which code in line 1382, code is update

File: cli/Cargo.toml (hunk 6 above)
Comment (baidang201) [reply cli/Cargo.toml:136]: yes, all features = [ "wasmtime" ] is seted in cli/Cargo.toml

File: runtime/src/lib.rs (hunk 4 above)
Comment (baidang201) [reply runtime/src/lib.rs:1638]: done!

**Review by AmadeusGB — DISMISSED** [review]

**Review by AmadeusGB — APPROVED** [review]

**Pull request merged**